### ISSUE TITLE
`SSL_get_ex_data_X509_STORE_CTX_idx` does not respect `OSSL_LIB_CTX`

### ISSUE THREAD

**Issue by matthauck** [openssl/openssl#24732]
The `SSL_get_ex_data_X509_STORE_CTX_idx()` function does not seem to respect the default `OSSL_LIB_CTX`. This can result in colliding ex-data indices when using custom library contexts, because `SSL_get_ex_data_X509_STORE_CTX_idx` is initialized statically, rather than once per library context. 

Repro with openssl 3.2.1:

```c++
#include <iostream>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/ssl.h>

int main( int argc, const char* argv[] )
{
    OPENSSL_add_all_algorithms_noconf();

    std::cout << "Main context:" << std::endl;
    std::cout << " - SSL_get_ex_data_X509_STORE_CTX_idx: " << SSL_get_ex_data_X509_STORE_CTX_idx() << std::endl;
    std::cout << " - next new index: "
              << X509_STORE_CTX_get_ex_new_index( 0, (void*)"name", nullptr, nullptr, nullptr ) << std::endl;

    OSSL_LIB_CTX_set0_default( OSSL_LIB_CTX_new() );

    std::cout << "New context:" << std::endl;
    std::cout << " - SSL_get_ex_data_X509_STORE_CTX_idx: " << SSL_get_ex_data_X509_STORE_CTX_idx() << std::endl;
    std::cout << " - next new index: " << X509_STORE_CTX_get_ex_new_index( 0, (void*)"name", nullptr, nullptr, nullptr )
              << std::endl;
}
```

Output:
```
Main context:
 - SSL_get_ex_data_X509_STORE_CTX_idx: 1
 - next new index: 2
New context:
 - SSL_get_ex_data_X509_STORE_CTX_idx: 1
 - next new index: 1
```

**Comment by nhorman** [comment]
I'm not sure I see the problem your describing.
SSL_get_ex_data_X509_STORE_CTX_idx isn't tracked per context, its tracked per process.  The idea behind this call and its family:

https://www.openssl.org/docs/man1.0.2/man3/X509_STORE_CTX_get_ex_new_index.html

Is to manage application 'slots' in the ex_data array for storing application specific data.  Its the responsibility of the application to record the return value from X509_STORE_CTX_get_ex_new_index so that it can be used later to store and fetch app specific data.

It arguably should be tracked per X509_STORE_CTX object, but currently its orthogonal to any LIB_CTX object

**Comment by matthauck** [comment]
The danger is that ex-data slots are unique per lib context.

https://github.com/openssl/openssl/blob/b76a6c26a254b4cc428275fc0ced56759dd5088a/crypto/context.c#L664-L670

As the test above demonstrates, you end up with `SSL_get_ex_data_X509_STORE_CTX_idx()` pointing to an index that overlaps with an index created via `X509_STORE_CTX_get_ex_new_index()` if you are using `OSSL_LIB_CTX_set0_default()`. This results in storing two different types of ex-data in the same slot, and applying the wrong/unexpected free/dup/etc. callbacks on that data.

**Comment by nhorman** [comment]
ok, thats a good point.  @mattcaswell can you please comment here?  This does seem problematic, but I'm not sure I see a good way to fix it without completely mangling the ABI here

**Comment by t8m** [comment]
Yeah, SSL_get_ex_data_X509_STORE_CTX_idx() should not use RUN_ONCE but store the index into the NULL libctx.
